### ISSUE TITLE
Windows port

### ISSUE THREAD

**Issue by froggynotacon** [brunoherbelin/vimix#37]
So I am very, very interested in seeing a Windows port of this.  I'm really rusty coder, so I don't feel confident enough setting out to do the port myself.  I would, however, be willing to join a team (if such a team exists) or assist with debugging and running test versions.  Since I searched and didn't see anything like that already I thought I'd ask.

Thanks!

**Comment by brunoherbelin** [comment]
Yeah, a windows port would actually be possible as vimix is based on multiplatform libraries.

This was evoked here https://github.com/brunoherbelin/vimix/issues/24 by @jorgerosa 
This could make 3 of us to try it: already a team :)

Although rusty, would you be able to install MS VisualStudio and pull source from gitlab?
Thanks

**Comment by froggynotacon** [comment]
I already have VisualStudio installed and have a basic familiarity with git.   I'm going on vacay this week but once I'm back I'll pull the source down, hit compile and see what breaks? :)

What I'm not necessarily comfortable with (after all these years) is how to properly check out and commit.  I've read so many git tutorials and I just don't get it.  I guess I'll need to find another :)

**Comment by brunoherbelin** [comment]
Thanks a lot!

Dependencies should not be too difficult as there are binaries for windows:
- gstreamer  https://gstreamer.freedesktop.org/download/
- glfw https://www.glfw.org/download.html 
- icu should be part of the OS https://docs.microsoft.com/en-us/dotnet/standard/globalization-localization/globalization-icu

When you clone the repository, don't forget to get other dependencies recursively ([README](https://github.com/brunoherbelin/vimix#readme)). 

Then, most OS specific changes should be only in one single file : https://github.com/brunoherbelin/vimix/blob/master/SystemToolkit.h

But in practice, i am sure it will require some adjustements. Let's see!

**Comment by Lorentz-Ipsum** [comment]
Hi @froggynotacon and @brunoherbelin
What is the state of this?
I did not see this open issue the other times I looked into this project.
I am also a bit of a rusty coder,  but I am looking forward to learn to make this possible since I see the lack of any good open source multiplatform VJ program and I would really love to improve my coding skills into a practical project like this.
I have no idea of what the general steps would be for porting the program to Windows,  so if anybody is open to meet with me and discuss how to approach this it would be great.
Until then,  I will look into details for the last message posted here and see what I can sort out on my own.

**Comment by brunoherbelin** [comment]
Hi!
Thanks for reviving this question! The issue is indeed still pending and I was not contacted or informed of other attempts at compiling for Windows.

I am very much willing to work with someone on this, but I simply do not have computers with Windows and no experience with MS Visual Studio and other Windows development environments. What I would need most is some help on setting up a compilation environment with the dependencies (glfw, gstreamer, etc.). 

I imagine that, if there was a VirtualBox image running a MSWindows system with the development environment setup, I could be able to contribute to the C++ code for making the functionalities Windows-compatible (i.e. in https://github.com/brunoherbelin/vimix/blob/master/SystemToolkit.h). Having this VBox would also allow to keep it up to date and compile updated packages.

Do you think this is something we could try?

**Comment by brunoherbelin** [comment]
Still looking for someone willing to help for the compilation of vimix under Windows!

**Comment by goatlady** [comment]
Huge disclaimer: I'm a web developer, not a Windows developer. I know just enough to be dangerous ;P

Microsoft make [VirtualBox VMs available for developers, Windows 11 with Visual Studio pre-installed](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/) (as well as WSL2 and Ubuntu). They are licensed for 60 days. 

I set one up and did the following:
- installed the C++ community tools/extensions (did not configure any settings)
- installed the gstreamer runtime and developer libraries
- downloaded glfw but wasn't sure what to do with it as there doesn't seem to be an installer - it's on the desktop
- cloned the git source in Visual Studio, which seemed to also download all the github-based dependencies but I'm not 100% sure how to check
- installed Firefox because life is too short for Edge

I'm packaging it up again as an .ova file. What's the best way to get it to you? Even compressed it will probably hit 20GB. I can share it on mega perhaps?

**Comment by brunoherbelin** [comment]
Excellent !!! Thanks a lot !
I have virtualbox and it should open .ova files.

Yes, a network transfer would be great ! You can share the download link here and I'll check ASAP.

**Comment by goatlady** [comment]
It's actually nearly 28GB, compressed to the max with 7zip, you can download it from here: 
https://mega.nz/file/NINyRIDI#-liraRStd-7Lr_kUhcHn5xblD-bJeuTNdAjSoiUB3tc

**Comment by brunoherbelin** [comment]
Got it !! Took ages :)  but got the archive and will test it during the week !

Thanks again for pointing me towards this MS Dev Environment. This is already a great help!

**Comment by goatlady** [comment]
Awesome, glad I could help 🤘

**Comment by goatlady** [comment]
[disclaimer: I'm not a C++ developer] It's possible that gstreamer needs to be added to the environment variables/system path. This might be helpful? https://dragonflycv.com/support/documentation/install-dragonfly/install-on-windows-10/

I played my first show doing live video mixing last Saturday with GLMixer and I'm hooked! But also super intrigued about the improvements that vimix has over GLMixer.

**Comment by j0r15** [comment]
Could you please update the link? The megadrive upload was expired. I'm no MS developer and can't get visual studio to work.

**Comment by brunoherbelin** [comment]
Hey [j0r15](https://github.com/j0r15)! Indeed the Win10 virtual box file created by [goatlady](https://github.com/goatlady) has expired.

It's not necessary however, and creating a virtual box of Window10 or 11 with either

- [Atlas](https://docs.atlasos.net/getting-started/installation/#prerequisites)
- [Microsoft virtual machines](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/)

is relatively straightforward using the downloadable disk image provided. 

From there, the real fun starts with the installation of the compilation environment and dependencies; cmake, gstreamer, etc. I am not familiar with MS Windows development, thus the absence of vimix version for Windows. And thus the idea is that, if we would be able to setup a development environment with everything ready to compile, we could use this windows virtual box to try to change the code to make it compile for windows... In principle, all dependencies exist for windows, and it should be doable...

**Comment by paperManu** [comment]
Hi there, I saw this thread while looking around what is the state of Vimix, and it happens that I ported a software I maintain to Windows a few months ago. And also that I wrote an article regarding how I did it overall.

Hopefully this will help, here it is: https://lab148.xyz/blog/port-splash-windows/

**Comment by lc-guy** [comment]
@brunoherbelin I managed to get a basic environment going on windows; couldn't manage to compile however, due to X11 / etc errors.

Here's how :

- Install msys2: https://www.msys2.org/
- Open an UCRT64 terminal
- `pacman -S mingw-w64-ucrt-x86_64-toolchain base-devel mingw-w64-ucrt-x86_64-cmake mingw-w64-ucrt-x86_64-libpng mingw-w64-ucrt-x86_64-glfw mingw-w64-ucrt-x86_64-gstreamer mingw-w64-ucrt-x86_64-gst-plugins-base mingw-w64-ucrt-x86_64-gst-libav mingw-w64-ucrt-x86_64-icu mingw-w64-ucrt-x86_64-gtk3 mingw-w64-ucrt-x86_64-glm mingw-w64-ucrt-x86_64-stb mingw-w64-ucrt-x86_64-tinyxml2 git`
- Attempt building vimix normally:
```bash
git clone --recursive https://github.com/brunoherbelin/vimix.git`
mkdir vimix-build
cd vimix-build
cmake -DCMAKE_BUILD_TYPE=Release ../vimix
```

I got a first error where it couldn't find GTK; running this diff seemed to fix it:
```diff
index a402332..ee744ab 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -108,6 +108,8 @@ if(UNIX)
 elseif(WIN32)
     add_definitions(-DWIN32)
     add_definitions(-DMINGW32)
+    find_package(GTK 3.0 REQUIRED)
+    macro_log_feature(GTK_FOUND "GTK" "GTK cross-platform widget toolkit" "http://www.gtk.org" TRUE)
 endif()

 #####
```

Which clears that up, but then fails with:
```
CMake Error at src/CMakeLists.txt:162 (target_link_libraries):
  Target "vimix" links to:

    X11::X11

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```

I couldn't manage to go further than this, sadly.
